docs/dev/kube-apiserver-health-check.md

```diff
@@ -0,0 +1,70 @@
+
+Now let's walk through the events (in chronological order) that unfold when a `kube-apiserver` instance restarts:
+* E1: `T+0s`: `kube-apiserver` receives a TERM signal.
+* E2: `T+0s`: `/readyz` starts reporting `failure` to signal to the load balancer that a shut down is in progress.
+  * The apiserver will continue to accept new request(s).
+  * The apiserver waits for certain amount of time (configurable by `shutdown-delay-duration`) before it stops accepting new request(s).
+* E3: `T+70s`: `kube-apiserver` (the http server) stops listening:
+  * `/healthz` turns red.
+  * Default TCP health check probe on port `6443` will fail.
+  * Any new request forwarded to it will fail, most likely with a `connection refused` error.
+  * Existing request(s) in-flight are not cut off but are allowed to complete gracefully, but there is a timeout of `60s`.
+* E4: `T+70s+60s`: Any existing request(s) that are still in flight are terminated.
+* E5: `T+70s+60s`: The apiserver process exits.
+
+An important note to consider is that today the time difference between `E3` and `E2` is `70s`. This is known as
+`shutdown-delay-duration` and is configurable.
+```
+$ kubectl -n openshift-kube-apiserver get cm config -o json | jq -r '.data."config.yaml"' |
+  jq '.apiServerArguments."shutdown-delay-duration"'
+[
+  "70s"
+]
+```
+In future we will reduce `shutdown-delay-duration` to `30s`. So for the sake of this document let's assume that `E3 - E2` is `30s`.
+
+Given the above, we can infer the following:
+* The load balancer should use `/readyz` endpoint for `kube-apiserver` health check probe. It must NOT use `/healthz` or
+default TCP port probe.
+* The time taken by the load balancer (let's say `t` seconds) to deem a `kube-apiserver` instance unhealthy and take it
+out of service should not bleed into `E3`. So `E2 + t < E3` must be true so that no new request is forwarded to the instance at `E3` or later. In the worst case it should
+take `30s` (since `E2` triggers) for the load balancer to take the `kube-apiserver` instance out of service.
+
+Below is the health check configuration used by `aws` currently.
+
+```
+protocol: HTTPS
+path: /readyz
+port: 6443
+unhealthy threshold: 2
+timeout: 10s
+interval: 10s
+```
+
+Let's plot the events as it happens when a load balance determines that a `kube-apiserver` is unhealthy and 
+takes it out of service. This considers a worst case scenario:
+* P1: T+0s:  `/readyz` starts reporting red.
+* P2: T+10s: first probe initiated.
+* P3: T+20s: first probe times out.
+* P4: T+20s: second probe initiated. (each health check is independent and will be kicked off at every interval)
+* P5: T+30s: second probe times out.
+* P6: T+30s: load balancer takes the unhealthy instance out of service.
+
+We assume the following:
+* Each health check request is independent and lasts the entire interval.
+* The time it takes for the instance to respond does not affect the interval for the next health check.
```

Comment:
do we know from aws/gcp docs that this is really the case?

Reply:
this is true for aws, i copied them verbatim from aws doc.

Reply:
link the docs, to make this easy to check?  Seems like it's [the classic-LB docs][1], but [the installer uses][2] [network load balancers][3] (classic LBs are [`aws_elb`][4]).

[1]: https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/elb-healthchecks.html
[2]: https://github.com/openshift/installer/blob/ca754a3e87b21fd9b861c15bfb97036f0c7eddde/data/data/aws/vpc/master-elb.tf#L1
[3]: https://www.terraform.io/docs/providers/aws/r/lb.html
[4]: https://www.terraform.io/docs/providers/aws/r/elb.html

Reply:
I could not find a doc that describes this for network load balancer exclusively. I think the health check mechanics should be the same for classic, application and network LB. Maybe we can ask this question to our AWS account rep.

On the other hand, what we stipulate above must hold true for all health checks universally. Otherwise if we allow one interval to bleed into another then we don't have a deterministic "at most".